h5py/h5d.pyx

```diff
@@ -534,47 +541,64 @@ cdef class DatasetID(ObjectID):
             compute the raw data. So the default value of `0` means that all
             defined filters have been applied to the raw data.
 
+            If the `out` argument is not None, it must be a writeable
+            contiguous 1D array-like of bytes (e.g., `bytearray` or
+            `numpy.ndarray`) and large enough to contain the whole chunk.
+
             Feature requires: 1.10.2 HDF5
             """
-
             cdef hid_t dset_id
             cdef hid_t dxpl_id
-            cdef hid_t space_id = 0
+            cdef hid_t space_id
             cdef hsize_t *offset = NULL
-            cdef size_t data_size
             cdef int rank
             cdef uint32_t filters
-            cdef hsize_t read_chunk_nbytes
-            cdef char *data = NULL
-            cdef bytes ret
+            cdef hsize_t chunk_bytes, out_bytes
+            cdef int nb_offsets = len(offsets)
+            cdef void * chunk_buffer
 
             dset_id = self.id
             dxpl_id = pdefault(dxpl)
-            space_id = H5Dget_space(self.id)
+            space_id = H5Dget_space(dset_id)
+            if space_id == -1:  # H5I_INVALID_HID
+                raise RuntimeError("Cannot retrieve dataset space")
             rank = H5Sget_simple_extent_ndims(space_id)
+            H5Sclose(space_id)
 
-            if len(offsets) != rank:
-                raise TypeError("offset length (%d) must match dataset rank (%d)" % (len(offsets), rank))
+            if nb_offsets != rank:
+                raise ValueError(
+                    f"offsets length ({nb_offsets}) must match dataset rank ({rank})"
+                )
+
+            offset = <hsize_t*> PyMem_Malloc(sizeof(hsize_t)*rank)
+            if not offset:
+                raise MemoryError()
```

Comment:
We mostly use our internal `emalloc` function, which deals with raising MemoryError. But I don't really understand when we need to use the Python malloc wrapper (as you've done) or the HDF5 one (`H5allocate_memory`). :confused:

Reply:
A bit of reading suggests that what's important is that it's the same memory management system allocating and freeing the memory. If that's how you understand it too, then I'd suggest we stick to our internal `emalloc` utility function for consistency, since we're in full control of where this is allocated and freed.

Reply:
I updated it to use `emalloc` as the other methods.

It was to use Python memory heap rather than system `malloc` as suggested in cython's [memory_allocation tutorial](https://cython.readthedocs.io/en/stable/src/tutorial/memory_allocation.html):

> Note that the C-API functions for allocating memory on the Python heap are generally preferred over the low-level C functions above as the memory they provide is actually accounted for in Python’s internal memory management system. They also have special optimisations for smaller memory blocks, which speeds up their allocation by avoiding costly operating system calls.

Reply:
Thanks. Maybe we do want to use that, but if so, let's do it in a separate PR. :+1: